### DISCUSSION TITLE
classes/class_boneattachment3d

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#282] [User-contributed notes]
# classes/class_boneattachment3d

Inherits: Node3D< Node< Object А node that dynamically copies or overrides the 3D transform of a bone in its parent Skeleton3D. Description: This node selects a bone in a Skeleton3D and attaches to...

https://docs.godotengine.org/en/stable/classes/class_boneattachment3d.html

**Comment by tancop** [comment]
This node doesn't respect transformations (move, rotate, scale) you do in the editor and always resets to the bone's transform at runtime. If you want to move/rotate the attached node relative to the skeleton, add a *child node* and move it relative to the `BoneAttachment`.

This tool script syncs the `BoneAttachment`'s transform to its bone so you can see where it actually ends up when you run the game:
```gdscript
@tool
extends BoneAttachment3D

func _process(_delta: float) -> void:
	if Engine.is_editor_hint() and not override_pose:
		var skel: Skeleton3D
		if get_use_external_skeleton():
			skel = get_node(get_external_skeleton())
		else:
			skel = get_parent()
			if skel is Skeleton3D and bone_idx < skel.get_bone_count():
				var xform: Transform3D = skel.get_bone_global_pose(bone_idx)
				global_transform = skel.global_transform * xform
```

**Comment by Acanixz** [comment]
Note: All ancestors of a BoneAttachment3D must inherit from Node3D, including the entire hierarchy from the scene root to the attachment.

If any parent in this chain is a plain Node, transform propagation will break. A common symptom is the attachment stopping its bone tracking when a non-looping animation finishes, only resuming once another animation starts.
